src/foam/util/NUIDGenerator.js

```diff
@@ -118,6 +115,31 @@ foam.CLASS({
             "NUIDGenerator: not supported on " + getSalt() + " without id property");
         }
       `
+    },
+    {
+      name: 'maybeUpdateSeqNo',
+      args: 'Long id',
+      javaCode: `
+        if ( id > 0x1000000 ) {
+          if ( UIDSupport.hash(id) != getHashKey() ) {
+            Loggers.logger(getX(), this).warning(getSalt(), "id:" + id, "hash not matched");
+            return;
+          }
+
+          var hex   = undoPermute(Long.toHexString(id));
+          var mid   = Integer.parseInt(hex.substring(hex.length() - 5, hex.length() - 3), 16);
+          if ( getMachineId() % 0xff == mid ) {
+            var seqNo = Integer.parseInt(hex.substring(0, hex.length() - 5), 16);
+            if ( seqNo > seqNo_.get() ) {
```

Comment:
All looks correct, but I think you could remove the double check and the synchronization by using the AtomicInt:  accumulateAndGet(int x, IntBinaryOperator accumulatorFunction) method.

Reply:
Fixed in https://github.com/kgrgreer/foam3/pull/1369.